### PR TITLE
Intersect the dirty rect with the tile

### PR DESCRIPTION
**Pull request by jrmuizel**
This prevents the dirty rect from extending beyond the size of tile
and causing explosions later on.

This fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1453718

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2653)

### CODE REVIEW

File: webrender/src/resource_cache.rs [hunk 1]
```diff
@@ -935,6 +936,20 @@ impl ResourceCache {
                 let (actual_width, actual_height) =
                     compute_tile_size(image_descriptor, tile_size, tile);
 
+                if let Some(dirty) = dirty_rect {
+                    dirty_rect = dirty
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: `dirty_rect` being `None` means that the whole image needs to be updated. In this case though, if there is zero intersection with the tile, we don't want any updates at all. I.e.:
```rust
if let Some(ref mut dirty) = dirty_rect {
  *dirty = match dirty.intersection(...) {
    Some(r) => {...; r}
    None => continue,
  };
}
```

File: webrender/src/resource_cache.rs [hunk 2]
```diff
@@ -935,6 +936,20 @@ impl ResourceCache {
...
+                        .intersection(&DeviceUintRect::new(
+                            DeviceUintPoint::new(tile.x as u32 * tile_size as u32, tile.y as u32 * tile_size as u32),
+                            DeviceUintSize::new(actual_width, actual_height)))
+                        .map(|mut x| {
+                            // we can't translate by a negate size so do it manually
+                            x.origin.x -= tile.x as u32 * tile_size as u32;
+                            x.origin.y -= tile.y as u32 * tile_size as u32;
+                            x
+                        }
+                        )
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: formatting

File: webrender/src/resource_cache.rs [hunk 3]
```diff
@@ -935,6 +936,24 @@ impl ResourceCache {
                 let (actual_width, actual_height) =
                     compute_tile_size(image_descriptor, tile_size, tile);
 
+                if let Some(ref mut dirty) = dirty_rect {
+                    *dirty = match dirty.intersection(&DeviceUintRect::new(
+                            DeviceUintPoint::new(
+                                tile.x as u32 * tile_size as u32,
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: I suppose this should be `(tile.x * tile_size) as u32`?

File: webrender/src/resource_cache.rs [hunk 4]
```diff
@@ -935,6 +936,24 @@ impl ResourceCache {
...
+                                tile.y as u32 * tile_size as u32
+                            ),
+                            DeviceUintSize::new(actual_width, actual_height),
+                    )) {
+                        Some(mut r) => {
+                            // we can't translate by a negative size so do it manually
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: we could probably have a `Vector2D` for that offset?

Comment (jrmuizel) [reply webrender/src/resource_cache.rs]: DeviceUintVector2D would still need positive values so I don't think that would work.

File: webrender/src/resource_cache.rs (hunk 3 above)
Comment (jrmuizel) [reply webrender/src/resource_cache.rs]: that would be u16 * u16 which we don't want.

Comment (kvark) [reply webrender/src/resource_cache.rs]: ah, I assumed those were floats.

**Comment by kvark** [comment]
@jrmuizel ok, I think this is fine to land now (r=me), unless you want to ship the raw test as well here.

**Comment by jrmuizel** [comment]
It looks like we might need to clip in two places. This patch does that, but I'll try to separate it out into a function.

**Comment by jrmuizel** [comment]
This should be ready now. I'll submit the test changes in follow ups.

**Pull request merged**